phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java

```diff
@@ -672,15 +927,15 @@ public void preBatchMutateWithExceptions(ObserverContext<RegionCoprocessorEnviro
       else {
           now = EnvironmentEdgeManager.currentTimeMillis();
       }
-      // First group all the updates for a single row into a single update to be processed
-      Collection<? extends Mutation> mutations = groupMutations(miniBatchOp, now, replayWrite);
+      // First merge all the updates for a single row into a single update to be processed
+      Collection<? extends Mutation> pendingMutations = mergeMutations(c, miniBatchOp, context, now, context.rebuild);
       // early exit if it turns out we don't have any edits
-      if (mutations == null) {
+      if (pendingMutations == null || pendingMutations.isEmpty()) {
           return;
       }
 
       long start = EnvironmentEdgeManager.currentTimeMillis();
-      prepareIndexMutations(c, miniBatchOp, context, mutations, now, indexMetaData);
+      preparePreIndexMutations(c, miniBatchOp, context, pendingMutations, now, indexMetaData);
```

Comment:
comment specifying that this is preparing the phase 1 unverified index writes would be helpful. Otherwise this sounds like you're preparing mutations for before the index write, which is a bit confusing.

Reply:
It's also  a little confusing that preparePreIndexMutations calls prepareIndexMutations. :-)

Reply:
ok